### PR TITLE
Asincify binding

### PR DESCRIPTION
**Pull request by LPeter1997**
Throws out untyped tree entirely by making binding out of order.

### CODE REVIEW

File: src/Draco.Compiler/Internal/Binding/ForLoopBinder.cs [hunk 1]
```diff
@@ -26,9 +26,10 @@ public ForLoopBinder(Binder parent, ForExpressionSyntax declaringSyntax)
         : base(parent, declaringSyntax)
     {
         this.DeclaringSyntax = declaringSyntax;
-        this.Iterator = new(
+        this.Iterator = new SourceLocalSymbol(
             this.ContainingSymbol!,
             declaringSyntax.Iterator.Text,
+            type: new TypeVariable(0), // TODO: Index
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Binding/ForLoopBinder.cs:32]: Solve TypeVariable construction

File: src/Draco.Compiler/Internal/Solver/ConstraintSolver.cs [hunk 2]
```diff
@@ -142,48 +122,30 @@ public LocalSymbol GetTypedLocal(UntypedLocalSymbol local, DiagnosticBag diagnos
     /// <returns>A new, unique type-variable.</returns>
     public TypeVariable AllocateTypeVariable(bool track = true)
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Solver/ConstraintSolver.cs:123]: Check, if every TV is constructed here, and the ones not need no tracking.

File: src/Draco.Compiler/Internal/Solver/ConstraintSolver.cs [hunk 3]
```diff
@@ -31,19 +40,23 @@ internal sealed partial class ConstraintSolver
...
         this.ContextName = contextName;
     }
 
+    // TODO: Docs
+    public Argument Arg(SyntaxNode? syntax, BindingTask<BoundExpression> expression, DiagnosticBag diagnostics) =>
+        new(syntax, expression.GetResultType(syntax, this, diagnostics));
+
+    // TODO: Docs
+    public Argument Arg(SyntaxNode? syntax, BindingTask<BoundLvalue> lvalue, DiagnosticBag diagnostics) =>
+        new(syntax, lvalue.GetResultType(syntax, this, diagnostics));
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Solver/ConstraintSolver.cs:58]: Document these

**Review by Kuinox — COMMENTED** [review]
Didn't had the patience to review everything...

File: src/Draco.Compiler/Internal/Binding/Binder.cs [hunk 4]
```diff
@@ -92,22 +91,27 @@ public virtual (TypeSymbol Type, BoundExpression? Value) BindGlobal(SourceGlobal
...
+                valueTask.GetResultType(valueSyntax, constraints, diagnostics),
+                global.DeclaringSyntax.Value!.Value);
         }
 
         // Solve
         constraints.Solve(diagnostics);
 
         // Type out the expression, if needed
-        var boundValue = untypedValue is null ? null : this.TypeExpression(untypedValue, constraints, diagnostics);
+        var boundValue = valueTask?.Result;
```
[the remark this answers is not in the document]
Comment (LPeter1997) [reply src/Draco.Compiler/Internal/Binding/Binder.cs:114]: The solver has long had the ability to detect this scenario and handles it by resolving the tasks to sentinel placeholders.
<https://github.com/Draco-lang/Compiler/blob/b726c017b5d35e4a3ee9bac516cedc93905ed6ce/src/Draco.Compiler/Internal/Solver/ConstraintSolver.cs#L119>

**Pull request merged**